src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,383 @@
+                parameters (torch.Tensor).
+            set_grad (bool): Whether to enable gradient calculation or not.
+
+        Returns:
+            torch.Tensor: Calculated mean value of KL divergence.
+
+        """
+        theta = dict(self.policy.named_parameters())
+        old_theta = dict(self._old_policy.named_parameters())
+
+        kls = []
+        for task_samples, task_params in zip(all_samples, all_params):
+            for i in range(self._num_grad_updates):
+                self._adapt(itr, task_samples[i], set_grad=set_grad)
+
+            update_module_params(self._old_policy, task_params)
+            with torch.set_grad_enabled(set_grad):
+                kl = self._inner_algo.compute_kl_constraint(
+                    task_samples[-1].observations)
+            kls.append(kl)
+
+            update_module_params(self.policy, theta)
+            update_module_params(self._old_policy, old_theta)
+
+        return torch.stack(kls).mean()
+
+    def _compute_loss(self, itr, batch_samples):
+        """Compute loss for one task and one gradient step.
+
+        Args:
+            itr (int): Iteration number.
+            batch_samples (garage.TorchTrajectoryBatch): Samples data for one
+                task and one gradient step.
+
+        Returns:
+            torch.Tensor: Computed loss value.
+
+        """
+        return self._inner_algo.compute_loss(itr, batch_samples)
+
+    def _compute_policy_entropy(self, task_samples):
+        """Compute policy entropy.
+
+        Args:
+            task_samples (list[garage.TorchTrajectoryBatch]): Samples data for
+                one task.
+
+        Returns:
+            torch.Tensor: Computed entropy value.
+
+        """
+        entropies = [
+            self._inner_algo.compute_policy_entropy(samples.observations)
+            for samples in task_samples
+        ]
+        return torch.stack(entropies).mean()
+
+    def _set_task(self, runner, task):
+        for env in runner._sampler._vec_env.envs:
+            env.set_task(task)
+
+    @property
+    def _old_policy(self):
+        """Old policy of inner algorithm.
+
+        Returns:
+            garage.torch.policies.base: Old policy of inner algorithm.
+
+        """
+        return self._inner_algo._old_policy
+
+    def _process_samples(self, itr, paths):
+        """Process sample data based on the collected paths.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths
+
+        Returns:
+            garage.TorchTrajectoryBatch: Processed samples data.
+
+        """
+        for path in paths:
+            path['returns'] = tensor_utils.discount_cumsum(
+                path['rewards'], self._inner_algo.discount)
+        self.baseline.fit(paths)
+        return self._inner_algo.process_samples(itr, paths)
+
+    def _log(self, itr, all_samples, loss_before, loss_after, kl_before, kl,
```

Comment:
i think this needlessly obfuscates logging, which is most useful in-line with code. logs are comments which happen to produce output.

i think you can split this into (1) a few well-place logger calls throughout the code and (2) one or more calls to a uniform evaluation function in RLAlgorithm (consult Zequn for info on (2)). If Zequn is not done by the time you commit this, you can include the evaluation part here, but he will have to refactor.